File: ProblemSet/912.sort-an-array/java/912.sort-an-array 快速排序 双向扫描 法1.java
import java.util.Random;

/*
 * @lc app=leetcode id=912 lang=java
 *
 * [912] Sort an Array
 */

// @lc code=start
class Solution {
    public int[] sortArray(int[] nums) {
        quickSort(nums, 0, nums.length - 1);
        return nums;
    }

    private void quickSort(int[] nums, int left, int right) {
        if (left < right) {
            int pos = partition(nums, left, right);
            quickSort(nums, left, pos - 1);
            quickSort(nums, pos + 1, right);
        }
    }

    private int partition(int[] nums, int left, int right) {
        int pivotIndex = new Random().nextInt(right - left + 1) + left;
        swap(nums, left, pivotIndex);
        int low = left + 1, high = right, pivot = nums[left];
        while (true) {
            while (low <= high && nums[low] < pivot) ++low;
            while (low <= high && nums[high] > pivot) --high;
            if (low < high) {
                swap(nums, low, high);
                ++low;
                --high;
            }
            else break;
        }
        swap(nums, left, high);
        return high;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

}
// @lc code=end

File: ProblemSet/912.sort-an-array/python/912.sort-an-array 快速排序 双向扫描 法1.py
import random

#
# @lc app=leetcode id=912 lang=python3
#
# [912] Sort an Array
#

# @lc code=start
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        self.quickSort(nums, 0, len(nums) - 1)
        return nums


    def quickSort(self, nums: List[int], left: int, right: int) -> List[int]:
        if left < right:
            pos = self.partition(nums, left, right)
            self.quickSort(nums, left, pos - 1)
            self.quickSort(nums, pos + 1, right)

    
    def partition(self, nums: List[int], left: int, right: int) -> int:
        pivotIdx = random.randint(0, right - left) + left
        nums[left], nums[pivotIdx] = nums[pivotIdx], nums[left]
        low, high, pivot = left + 1, right, nums[left]
        while True:
            while low <= high and nums[low] < pivot:
                low += 1
            while low <= high and nums[high] > pivot:
                high -= 1
            if low < high:
                nums[low], nums[high] = nums[high], nums[low]
                low += 1
                high -= 1
            else:
                break
        nums[left], nums[high] = nums[high], nums[left]
        return high


# @lc code=end

